src/extension/wmts-multi-dimensional/src/main/java/org/geoserver/gwc/wmts/dimensions/DimensionsUtils.java

```diff
@@ -246,32 +261,166 @@ static <T> Tuple<T, T> getMinMax(List<Comparable> values, Class<T> type) {
         return uniqueVisitor.getUnique();
     }
 
+    private static Set<Comparable> resultToComparableSet(
+            CalcResult result, String endAttributeName, SortBy sortBy) {
+        if (result == CalcResult.NULL_RESULT) return new HashSet<>();
+        @SuppressWarnings("unchecked")
+        Set<Object> values = result.toSet();
+        if (values.isEmpty()) return Collections.emptySet();
+        else return toComparableSet(values, endAttributeName, sortBy);
+    }
+
+    private static Set<Comparable> toComparableSet(
+            Set<Object> set, String endAttribute, SortBy sortBy) {
+        Set<Comparable> resultSet = getTreeSet(sortBy, endAttribute);
+        Iterator<Object> objects = set.iterator();
+
+        while (objects.hasNext()) {
+            Object val = objects.next();
+            @SuppressWarnings("unchecked")
+            List<Object> values = (List<Object>) val;
+            Object start = values.get(0);
+            Object end = values.get(1);
+            Comparable range = (Comparable) toRange(start, end);
+            if (range != null) resultSet.add(range);
+        }
+        return resultSet;
+    }
+
+    private static TreeSet<Comparable> getTreeSet(SortBy sortBy, String endAttribute) {
+        TreeSet<Comparable> resultSet = new TreeSet<>();
+        if (sortBy != null) {
+            if (isSortByEnd(endAttribute, sortBy)) {
+                Comparator<Comparable> comparator = comparatorSortByEnd();
+                if (sortBy.getSortOrder().equals(SortOrder.DESCENDING))
+                    comparator = comparatorSortByEnd().reversed();
+                resultSet = new TreeSet<>(comparator);
+            } else if (sortBy.getSortOrder().equals(SortOrder.DESCENDING)) {
+                resultSet = new TreeSet<>(comparatorSortByStart().reversed());
+            }
+        }
+        return resultSet;
+    }
+
+    private static boolean isSortByEnd(String endAttribute, SortBy sortBy) {
+        if (endAttribute == null) return false;
+        return sortBy.getPropertyName().getPropertyName().equals(endAttribute);
+    }
+
+    private static Comparator<Comparable> comparatorSortByEnd() {
+        return new Comparator<Comparable>() {
+            @Override
+            @SuppressWarnings("unchecked")
+            public int compare(Comparable o1, Comparable o2) {
+                if (o1 instanceof ComparableRange && o2 instanceof ComparableRange) {
+                    ComparableRange comp1 = (ComparableRange) o1;
+                    ComparableRange comp2 = (ComparableRange) o2;
+                    int result = comp1.getMaxValue().compareTo(comp2.getMaxValue());
+                    if (result == 0) result = comp1.getMinValue().compareTo(comp2.getMinValue());
+                    return result;
+                }
+                return o1.compareTo(o2);
+            }
+        };
+    }
+
+    private static Comparator<Comparable> comparatorSortByStart() {
```

Comment:
This and the previous method should really be inner classes, e.g:

```java
private static Class ComparatorSortByStart implements Comparator<Comparable> { ... }
```
and possibly have a single instance in a constant, rather than being created over and over (there is nothing dynaminc in them).

Reply:
fixed